h/services/flag.py

```diff
@@ -7,16 +7,52 @@ class FlagService:
     def __init__(self, session):
         self._session = session
 
+        self._flagged_cache = {}
+
+    def create(self, user: User, annotation: Annotation):
+        """
+        Create a flag for the given user and annotation.
+
+        We enforce the uniqueness of a flag, meaning one user can only
+        flag one annotation once. This method first checks if the annotation
+        is already flagged by the user, if that is the case, then this
+        is a no-op.
+
+        :param user: The user flagging the annotation.
+        :param annotation: The annotation to be flagged.
+        """
+        if self.flagged(user, annotation):
+            return
+
+        self._session.add(Flag(user=user, annotation=annotation))
+
     def flagged(self, user: User, annotation: Annotation):
         """
         Check if a given user has flagged a given annotation.
 
+        You can make this more efficient for a large batch of annotations by
+        calling `all_flagged()` which will prime a cache of results.
+
         :param user: The user to check for a flag.
         :param annotation: The annotation to check for a flag.
         :returns: True/False depending on the existence of a flag.
         """
-        query = self._session.query(Flag).filter_by(user=user, annotation=annotation)
-        return query.count() > 0
+        if not user or not annotation:
+            return False
+
+        # This cache can be primed by calling `all_flagged()`
+        key = user.id, annotation.id
+        if key in self._flagged_cache:
+            return self._flagged_cache[key]
+
+        is_flagged = bool(
+            self._session.query(Flag)
+            .filter_by(user=user, annotation=annotation)
+            .one_or_none()
```

Comment:
This will raise if there's more than one Flag, but not sure if that's a possibility

Reply:
Oh god, that's absolutely a possibility.

Reply:
Ok, so this is actually fine. There's a unique constraint between users and annotations for flagging, so there can be at most one of these.